core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -54,17 +72,13 @@ public TypeAnalyzer(PlannerContext plannerContext, StatementAnalyzerFactory stat
 
     public Map<NodeRef<Expression>, Type> getTypes(Session session, TypeProvider inputTypes, Iterable<Expression> expressions)
     {
-        return analyzeExpressions(
-                session,
-                plannerContext,
-                statementAnalyzerFactory,
-                new AllowAllAccessControl(),
-                inputTypes,
-                expressions,
-                ImmutableMap.of(),
-                WarningCollector.NOOP,
-                OTHERS)
-                .getExpressionTypes();
+        try {
+            ExpressionAnalyzer analyzer = analyzerCache.get(new CacheIdentityKey(getIdFromSession(session), expressions), () -> createExpressionAnalyzer(session, plannerContext, statementAnalyzerFactory, inputTypes));
+            return analyzeExpressions(analyzer, expressions).getExpressionTypes();
```

Comment:
Just call `ExpressionAnalyzer#analyze` directly, no need to go through `ExpressionAnalysis`